modules/otr.cpp

```diff
@@ -0,0 +1,1291 @@
+            "conversation. However, you do not have a plugin to support that. "
+            "See http://otr.cypherpunks.ca/ for more information.";
+    }
+
+    bool TargetIsChan(const CString& sTarget) {
+        CIRCNetwork* network = GetNetwork();
+        assert(network);
+
+        if (sTarget.empty()) {
+            return true;
+        } else if (network->GetChanPrefixes().empty()) {
+            // RFC 2811
+            return (CString("&#!+").find(sTarget[0]) != CString::npos);
+        } else {
+            return (network->GetChanPrefixes().find(sTarget[0]) !=
+                    CString::npos);
+        }
+    }
+
+    EModRet SendEncrypted(CString& sTarget, CString& sMessage) {
+        gcry_error_t err;
+        char* newmessage = NULL;
+        const char* accountname = GetUser()->GetUserName().c_str();
+        CString sNick = sTarget.AsLower();
+
+        inject_workaround_mod = this;
+        err = otrl_message_sending(
+            m_pUserState, &m_xOtrOps, this, accountname, PROTOCOL_ID,
+            sNick.c_str(), OTRL_INSTAG_BEST, sMessage.c_str(), NULL,
+            &newmessage, OTRL_FRAGMENT_SEND_ALL, NULL, COtrAppData::Add, NULL);
+        inject_workaround_mod = NULL;
+
+        if (err) {
+            PutModuleBuffered(CString("otrl_message_sending failed: ") +
+                              gcry_strerror(err));
+            return HALT;
+        }
+
+        if (newmessage) {
+            // libotr injected the message
+            otrl_message_free(newmessage);
+            return HALT;
+        } else {
+            // not an OTR message
+            return CONTINUE;
+        }
+    }
+
+    virtual EModRet OnUserMsg(CString& sTarget, CString& sMessage) {
+        // Do not pass the message to libotr if sTarget is a channel
+        if (TargetIsChan(sTarget) || IsIgnored(sTarget)) {
+            return CONTINUE;
+        }
+
+        return SendEncrypted(sTarget, sMessage);
+    }
+
+    virtual EModRet OnUserAction(CString& sTarget, CString& sMessage) {
+        if (TargetIsChan(sTarget) || IsIgnored(sTarget)) {
+            return CONTINUE;
+        }
+
+        // http://www.cypherpunks.ca/pipermail/otr-dev/2012-December/001520.html
+        // suggests using following:
+        // CString sLine = "\001ACTION " + sMessage + "\001";
+        // However, irssi and weechat plugins send it like this:
+        CString sLine = "/me " + sMessage;
+
+        // Try sending the formatted line. If CONTINUE is returned,
+        // pass unformatted to other plugins/caller.
+        return SendEncrypted(sTarget, sLine);
+    }
+
+    virtual EModRet OnPrivMsg(CNick& Nick, CString& sMessage) {
+        int res;
+        char* newmessage = NULL;
+        OtrlTLV* tlvs = NULL;
+        ConnContext* ctx = NULL;
+        CString sNick = Nick.GetNick().AsLower();
+
+        if (IsIgnored(Nick.GetNick())) {
+            return CONTINUE;
+        }
+
+        const char* accountname = GetUser()->GetUserName().c_str();
```

Comment:
Shouldn't be a problem, each network has its own nick (context) database. Also each network has its own private key:(